src/kvp.rs

```diff
@@ -0,0 +1,684 @@
+/// proper formatting for consumption by hv_kvp_daemon service,
+/// which typically reads from /var/lib/hyperv/.kvp_pool_1.
+fn encode_kvp_slice(key: Vec<u8>, value: Vec<u8>) -> Vec<u8> {
+    let mut buffer = Vec::with_capacity(
+        HV_KVP_EXCHANGE_MAX_KEY_SIZE + HV_KVP_EXCHANGE_MAX_VALUE_SIZE,
+    );
+    buffer.extend_from_slice(&key);
+    buffer.extend_from_slice(&value);
+    buffer
+}
+
+/// Decodes a KVP byte slice into its corresponding key and value strings.
+/// This is useful for inspecting or logging raw KVP data.
+#[cfg(test)]
+pub fn decode_kvp_item(
+    record_data: &[u8],
+) -> Result<(String, String), &'static str> {
+    let record_data_len = record_data.len();
+    let expected_len =
+        HV_KVP_EXCHANGE_MAX_KEY_SIZE + HV_KVP_EXCHANGE_MAX_VALUE_SIZE;
+
+    if record_data_len != expected_len {
+        return Err("record_data len not correct.");
+    }
+
+    let key = String::from_utf8(
+        record_data[0..HV_KVP_EXCHANGE_MAX_KEY_SIZE].to_vec(),
+    )
+    .unwrap_or_else(|_| String::new())
+    .trim_end_matches('\x00')
+    .to_string();
+
+    let value = String::from_utf8(
+        record_data[HV_KVP_EXCHANGE_MAX_KEY_SIZE..]
+            .iter()
+            .take(HV_KVP_AZURE_MAX_VALUE_SIZE)
+            .cloned()
+            .collect(),
+    )
+    .unwrap_or_else(|_| String::new())
+    .trim_end_matches('\x00')
+    .to_string();
+
+    Ok((key, value))
+}
+
+/// Truncates the guest pool KVP file if it contains stale data (i.e., data
+/// older than the system's boot time). Logs whether the file was truncated
+/// or no action was needed.
+fn truncate_guest_pool_file(kvp_file: &Path) -> Result<(), Error> {
+    let boot_time = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .map_err(|e| Error::new(std::io::ErrorKind::Other, e))?
+        .as_secs()
+        - get_uptime()?.as_secs();
+
+    match kvp_file.metadata() {
+        Ok(metadata) => {
+            if metadata.mtime() < boot_time as i64 {
+                OpenOptions::new()
+                    .write(true)
+                    .truncate(true)
+                    .open(kvp_file)?;
+                println!("Truncated the KVP file due to stale data.");
+            } else {
+                println!(
+                    "File has been truncated since boot, no action taken."
+                );
+            }
+        }
+        Err(ref e) if e.kind() == ErrorKind::NotFound => {
+            println!("File not found: {:?}", kvp_file);
+            return Ok(());
+        }
+        Err(e) => {
+            return Err(e);
+        }
+    }
+
+    Ok(())
+}
+
+/// Retrieves the system's uptime using the `sysinfo` crate, returning the duration
+/// since the system booted. This can be useful for time-based calculations or checks,
+/// such as determining whether data is stale or calculating the approximate boot time.
+fn get_uptime() -> Result<Duration, Error> {
```

Comment:
```suggestion
fn get_uptime() -> Duration {
```
I'm guessing this was from when you moved to `sysinfo`, but it looks like there's no calls in the function that are fallible so you can return the Duration directly.